### PR TITLE
Implement Tiling on non bmm Reduction Ranges for automatic hints

### LINKED ISSUE #3198: Implement Tiling on Reduction Ranges - for automatic hints
This task is to implement the tiling on Reduction Ranges for Reduction ops for automatic hints

### LINKED ISSUE #4281: Implement Tiling for Non Bmm Reduction cases
### What feature would you like to be added?

This feature will introduce to emit the automatic tiling pass for non bmm reduction cases

### Why is this feature needed?

to emit the automatic hints when a given non bmm  reduction case exceeds the span limit 

### Any alternative solution that you may want to share?

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Hemaprasannakc**
Resolves #3198 (sub-issue of #206 "Working set reduction").
Resolves #4281

### What this covers

Introduces tiling on reduction dimension of generic non bmm reduction ops . Ops covered  - "sum","prod","max","min"
Before this PR the **automatic** span-overflow pass could only do this forbatch-matmul K (#3662, #4000). A plain `sum` / `prod` / `max` / `min` whose *reduced axis* drove the overflow was silently skipped — the span stayed over the limit and surfaced much later as a confusing backend failure.

This PR extends the automatic pass to those four reduction types. It is a **planner + adapter** change only — everything downstream in `coarse_tile.py`(identity-fill + per-tile combine + drain) was already generic and is
untouched.

For any other non bmm reduction op overflows apart from ops covered above, raises a clear Unsupported at compile time (new  _has_untileable_reduction_span) instead of silently dropping the overflow.

### Also in this PR: un-xfail 4 stale span-overflow tests
Investigated the 4 @unittest.expectedFailure tests in this file with latest code and changed it

test_auto_span_overflow_matches_equivalent_spyre_hint_loop_spec — asserted the planner picks a hard-coded split 5; it validly picks 4. Rewritten to read the planner's chosen split, feed it to the manual spyre_hint, and compare LoopSpec blocks.

test_reduction_input_span_codegen_contains_auto_loop_spec — asserted the tiled loop symbol is literally c0; coarse tiling renamed it _tile_adv_op<N>_lvl<L>. Rewritten to assert structure (one tiled symbol + trip count).

test_bmm_to_pointwise_join_numeric / test_bmm_to_reduction_join_numeric — asserted a fixed atol/rtol on a reordered fp16 BMM. Verified tiled and untiled runs have identical mismatch counts vs CPU (the coarse tile adds zero error). The genuine past bugs were fixed by #3612 / #3613; the comment and decorator were never removed. Rewritten to the tiled-vs-untiled comparison already used by test_lm_head_matmul_join_numeric.

test_span_overflow_hint_analysis.py: 167 passed, 0 xfailed

### Not covered yet
mean reduction-range tiling (scale-factor handling) 
More than one reduction range tiled on a single op.
can_conform_reduction_tile — a Reduction consumer adopting a producer group's split in producer/consumer fusion — separate limitation, tracked in #3625.
Output N/Vocab tiling for non-divisible / prime sizes — separate feature, tracked in #3203.

### Docs
docs/source/compiler/span_overflow_hint_analysis.md and coarse_tiling_loops.md updated: support matrix, reduction flow, failure policy, known limitations, and the producer/consumer coverage matrix (now
all-pass — the stale prose about the 4 xfails removed).

### test logs

```
(dt-inductor) hemaprasanna@hemaprasanna-torch-spyre:~/dt-inductor/torch-spyre$ python3 -m pytest -q tests/inductor/test_span_overflow_hint_analysis.py
......................................................................................................................................................... [ 90%]
................                                                                                                                                              [100%]
========================================================================= warnings summary ==========================================================================
tests/inductor/test_span_overflow_hint_analysis.py: 31 warnings
  /home/hemaprasanna/dt-inductor/.venv/lib64/python3.12/site-packages/torch/_dynamo/pgo.py:605: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
    warn_once(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
169 passed, 31 warnings, 4 subtests passed in 26.29s
(dt-inductor) hemaprasanna@hemaprasanna-torch-spyre:~/dt-inductor/torch-spyre$ 
```

### CODE REVIEW

**Review by dgrove-oss — DISMISSED** [review]
Summary

Extends the automatic span-overflow coarse-tiling planner so plain sum/prod/max/min reductions (not just batch-matmul K) can have their single reduction range tiled, reusing coarse tiling's existing identity-fill + per-tile-combine + drain mechanism. Adds _has_untileable_reduction_span to raise a clear Unsupported at compile time instead of letting an unfixable overflow silently reach the backend, fixes a latent crash risk in reduction_loop_vars (a StarDep whose .index raises NotImplementedError, previously caught only by luck since hasattr doesn't guard against that), and un-xfails 4 stale tests whose original failures were already resolved by earlier PRs (#3612/#3613) but never cleaned up.

Review

BLOCKERs

None.

SUGGESTIONs

- torch_spyre/model_utils.py (+5/-3): a pure reformatting change (parenthesizing a conditional lambda), unrelated to this PR's stated purpose — likely picked up from a local pre-commit run --all-files. Consider keeping unrelated formatting fixes out of feature PRs.
- test_mean_reduction_range_overflow_raises_at_compile asserts on bare Exception with a message substring rather than the Unsupported type used by sibling unit tests — reasonable given Inductor's compile-time exception wrapping, but worth tightening if a stable wrapped type exists.

QUESTIONs

None. I investigated the 3 CI jobs that failed on both their original run and pod-level retry (Test Inductor Ops Lx Matmul, Test Inductor Ops Matmul, Test Memory Pressure Gc) by pulling raw job logs directly. All three are confirmed hardware/infra RAS errors unrelated to this PR's code:
- RAS::VFIO::DeviceOpenFail — "PCIe address was not found in the /dev/vfio subsystem... Device or resource busy" (device-open collision, twice, on two different jobs).
- RAS::RUNTIMESCHEDULER::ComputeHardwareError — "A hardware error was detected in the compute control block response," alongside a DeviceMemAllocFailed.

None of these touch matmul work-division or memory-pressure logic that this PR modifies. Run Spyre unit tests and Test matrix result are pure result-aggregator gate jobs with no independent logic — they fail solely because they propagate the above.

Notes on quality

- The core planner logic (_supports_reduction_range_tiling, _has_untileable_reduction_span, _best_case_inner_span, the widened _bmm_k_symbol) is unusually well-documented — every scope decision (BMM exemption, dtype restriction on max/min, why max_cores isn't consulted, why a mixed coordinate is charged only its reduction part) has an explicit rationale, and I independently verified the central claim against work_division_constraints._K_SPLIT_COMBINE_SUPPORTED: plain sum/prod/max/min are indeed absent from that set, pinning their reduction vars to allowed_splits={1} — confirming the hard-error path is genuinely necessary, not just defensive.
- reduction_loop_vars's isinstance(d, MemoryDep) fix (replacing hasattr(d, "index")) is a real, well-justified fix, not cosmetic.
- All 4 un-xfailed tests are genuine fixes (verified against actual before/after bodies), not loosened assertions.
- New test coverage under TestSpanOverflowGenericReductionRangeTiling is thorough: unit-level (including test_integer_max_min_not_reduction_range_tiled and test_untileable_check_skips_integer_reductions for the dtype restriction, and multi-reduction-range/kill-switch/StarDep-crash-guard edge cases), codegen (flat and nested output+reduction), and on-device numeric tests for sum/max/min/prod plus an "organic" test using the real planner's own choice rather than a forced plan.
- Both doc files accurately reflect the code changes and correctly resolve the historical "known limitation" prose, attributing prior xfails to already-merged fixes (#3612/#3613).

Verdict

APPROVE

File: torch_spyre/model_utils.py [hunk 1]
```diff
@@ -292,9 +292,11 @@ def _transfer_module(
     """
     if _module_overrides_apply(module):
         module._apply(
-            lambda t: _dma_to_spyre_default(t, target_dtype=dtype)
-            if t is not None and t.device.type != DEVICE_NAME
-            else t
+            lambda t: (
```
Comment (dgrove-oss) [inline torch_spyre/model_utils.py]: This reformat makes the PR require additional approvers.  Drop it?

Comment (Hemaprasannakc) [reply torch_spyre/model_utils.py]: dropped them

**Comment by Hemaprasannakc** [comment]
@dgrove-oss 

Thanks for the thorough review and for taking the time to pull the raw CI job logs to confirm those 3 failures are unrelated hardware/RAS infra issues, not code from this PR.

Both suggestions are addressed in `813e0e77`:

- **`model_utils.py` reformat** — dropped. It was pure ruff-formatting noise pulled in by an earlier `git add -A`, unrelated to this PR's purpose. The file is now byte-identical to `main`:

  ```
  $ git diff upstream/main -- torch_spyre/model_utils.py
  (no output)
  ```

- **`test_mean_reduction_range_overflow_raises_at_compile`** — tightened from bare `Exception` to `torch._inductor.exc.InductorError`. Verified that Inductor wraps the `Unsupported` error in `InductorError` at the compile boundary, and confirmed it's already the established pattern for this exact scenario elsewhere in the suite (`test_coarse_tile_e2e.py`, `test_restickify.py`, `test_codegen.py`, `test_fp8_operations.py`).

Full suite re-run after both changes:

```
$ python3 -m pytest tests/inductor/test_span_overflow_hint_analysis.py -q
........................................................................ [ 38%]
..................................................................... [ 75%]
.............................................                [100%]
186 passed, 15 subtests passed in 28.31s
```

```
$ python3 -m pytest tests/inductor/test_enumerate_tilings.py tests/inductor/test_coarse_tiling.py tests/tensor/test_model_utils.py -q
........................................................................ [ 17%]
................................................................ [ 33%]
........................................................................ [ 51%]
........................................................................ [ 69%]
........................................................................ [ 86%]
.....................................................                    [100%]
405 passed, 8 subtests passed in 8.01s
```

```
$ python3 -m pytest tests/inductor/test_coarse_tile_e2e.py -q -k "min_ or reduce_both or reduce_dim or add_min"
...........................                                              [100%]
27 passed, 210 deselected in 30.83s
```

**Review by dgrove-oss — APPROVED** [review]
PR #4448 Review: Implement Tiling on non-bmm Reduction Ranges for automatic hints

Summary

Extends the automatic span-overflow coarse-tiling planner so plain sum/prod/max/min reductions (not just batch-matmul K) can have their single reduction range tiled, reusing coarse tiling's existing identity-fill + per-tile-combine + drain mechanism. Adds _has_untileable_reduction_span to raise a clear Unsupported at compile time instead of letting an unfixable span overflow silently reach the backend, fixes a latent crash risk in reduction_loop_vars (a StarDep whose .index raises NotImplementedError, previously only accidentally avoided since hasattr doesn't guard against that), and un-xfails 4 stale tests whose original failures were already fixed by #3612/#3613 but never cleaned up.

BLOCKERs: None.

SUGGESTIONs:
- test_mean_reduction_range_overflow_raises_at_compile asserts on a bare Exception with a message substring rather than the Unsupported type sibling unit tests use — reasonable given Inductor's compile-time exception wrapping, but worth tightening later if a stable wrapped type exists. (non-blocking)

QUESTIONs: None. CI is green across the board.

Notes on quality:
- The new planner functions are unusually well-documented, with every scope decision independently verifiable against work_division_constraints._K_SPLIT_COMBINE_SUPPORTED.
- The description's claim that coarse_tile.py is "untouched" is slightly imprecise — reduction_loop_vars did change (a real, well-justified StarDep crash fix) — but the core identity-fill/combine/drain machinery is indeed untouched.
- All 4 un-xfailed tests are genuine improvements, not just decorator removal.
- Error messages follow the Unsupported(...) convention and are specific.

Verdict: APPROVE
